src/vs/workbench/contrib/notebook/browser/viewParts/notebookEditorStickyScroll.ts

```diff
@@ -0,0 +1,342 @@
+		this.renderStickyLines(trackedEntry?.parent, this._stickyScrollContent, linesToRender);
+		this.updateDisplay();
+	}
+
+	private updateContent() {
+		// find first code cell in visible range. this marks the start of the first section
+		// find the last code cell in the first section of the visible range, store the bottom scroll position in a const sectionBottom
+		// compute sticky scroll height, and check if editorScrolltop + stickyScrollHeight < sectionBottom // ? maybe use 22 instead of stickyscrollheight, as we don't necessarily render each line
+		// if that condition is true, break out of the loop with that cell as the tracked cell
+		// if that condition is false, continue to next cell
+
+		DOM.clearNode(this._stickyScrollContent);
+		const editorScrollTop = this.notebookEditor.scrollTop;
+
+		// find last code cell of section, store bottom scroll position in sectionBottom
+		const visibleRange = this.notebookEditor.visibleRanges[0];
+		if (!visibleRange) {
+			return;
+		}
+
+		let trackedEntry = undefined;
+		let sectionBottom = 0;
+		for (let i = visibleRange.start; i < visibleRange.end; i++) {
+			if (i === 0) { // don't show headers when you're viewing the top cell
+				this.updateDisplay();
+				return;
+			}
+			const cell = this.notebookEditor.cellAt(i);
+			if (!cell) {
+				return;
+			}
+			if (cell.cellKind === CellKind.Markup) {
+				continue;
+			}
+
+			// if we are here, the cell is a code cell.
+			// check next cell, if markdown, that means this is the end of the section
+			const nextCell = this.notebookEditor.cellAt(i + 1);
+			if (nextCell) {
+				if (nextCell.cellKind === CellKind.Markup) {
+					// this is the end of the section
+					// store the bottom scroll position of this cell
+					sectionBottom = this.notebookCellList.getCellViewScrollBottom(cell);
+					// compute sticky scroll height
+					const entry = this.getVisibleOutlineEntry(i);
+					if (!entry) {
+						return;
+					}
+					// check if we can render this section of sticky
+					const currentSectionStickyHeight = this.computeStickyHeight(entry!);
+					if (editorScrollTop + currentSectionStickyHeight < sectionBottom) {
+						const linesToRender = Math.floor((sectionBottom - editorScrollTop) / 22);
+						this.renderStickyLines(entry?.parent, this._stickyScrollContent, linesToRender);
+						break;
+					}
+
+					let nextSectionEntry = undefined;
+					for (let j = 1; j < visibleRange.end - i; j++) {
+						// find next code cell after this one
+						const cellCheck = this.notebookEditor.cellAt(i + j);
+						if (cellCheck && cellCheck.cellKind === CellKind.Code) {
+							nextSectionEntry = this.getVisibleOutlineEntry(i + j);
+							break;
+						}
+					}
+					const nextSectionStickyHeight = this.computeStickyHeight(nextSectionEntry!);
+
+
+					// todo: this logic needs cleaning... bad. but it works for now.
+					// ? lines to render clearly needs a rethink, shouldn't be adding 100 to override it
```

Comment:
We might want to have comments for this section of code, explaining how each `if` clause works. Also I'm not sure how it works if we have h1 all the way to h6.